### PR TITLE
calculateMDS/runMDS accept custom distance function

### PR DESCRIPTION
**Pull request by FelixErnst**

### CODE REVIEW

File: R/runMDS.R [hunk 1]
```diff
@@ -11,23 +11,26 @@
 #' @param ... For the \code{calculateMDS} generic, additional arguments to pass to specific methods.
 #' For the SummarizedExperiment and SingleCellExperiment methods, additional arguments to pass to the ANY method.
 #'
-#' For \code{runMDS}, additional arguments to pass to \code{calculateMDS}. 
-#' @param method String specifying the type of distance to be computed between cells.
+#' For \code{runMDS}, additional arguments to pass to \code{calculateMDS}.
+#' @param FUN a function returning a \code{dist} object from a matrix, where rows are samples and rows are features
```
Comment (LTLA) [inline R/runMDS.R]: Capital `A`, end with a period.

File: R/runMDS.R [hunk 2]
```diff
@@ -11,23 +11,26 @@
...
+#' @param keep_dist \code{TRUE} or \code{FALSE}: should the \code{dist} calculated by \code{FUN}
+#'   object be stored as \sQuote{dist} attribute of the matrix returned/stored by \code{calculateMDS}/
+#'   \code{runMDS}? (default: \code{keep_dist = FALSE})
```
Comment (LTLA) [inline R/runMDS.R]: `Logical scalar indicating whether the \code{dist}`... No question mark either.

No need to specify the default, that's what the `usage` is for.

File: R/runMDS.R [hunk 3]
```diff
@@ -44,15 +47,23 @@
 NULL
 
 #' @importFrom stats cmdscale dist
-.calculate_mds <- function(x, ncomponents = 2, ntop = 500, subset_row = NULL, 
-    scale=FALSE, transposed=FALSE, method = "euclidean")
+.calculate_mds <- function(x, FUN = stats::dist, ncomponents = 2,
```
Comment (LTLA) [inline R/runMDS.R]: Is there a non-**stats** `dist` you're concerned about?

File: R/runMDS.R [hunk 4]
```diff
@@ -44,15 +47,23 @@
...
+        stop("'keep_dist' must be TRUE or FALSE.", call. = FALSE)
+    }
     if (!transposed) {
-        x <- .get_mat_for_reddim(x, subset_row=subset_row, ntop=ntop, scale=scale) 
+        x <- .get_mat_for_reddim(x, subset_row=subset_row, ntop=ntop, scale=scale)
+    }
+    x <- as.matrix(x)
+    cell_dist <- do.call(FUN, c(list(x),list(...)))
+    ans <- cmdscale(cell_dist, k = ncomponents)
+    if(keep_dist){
```
Comment (alanocallaghan) [inline R/runMDS.R]: `if (x) {` rather than `if(x){` please, and spaces after commas

File: R/runMDS.R (hunk 3 above)
Comment (FelixErnst) [reply R/runMDS.R]: Not really, but I wouldn't bet on it.

Comment (LTLA) [reply R/runMDS.R]: Package namespacing should protect you against other symbols in the search path.

File: R/runMDS.R (hunk 1 above)
Comment (LTLA) [reply R/runMDS.R]: Also, it seems that `FUN` just needs to return a distance matrix, not necessary a `dist` object.

I don't really understand what you mean by "rows are samples and rows are features". You should say something like:

> A function that accepts a numeric matrix as its first argument, where rows are samples and columns are features; and returns a numeric distance matrix (or an object coercible into one) of order equal to the number of samples.

**Comment by FelixErnst** [comment]
Hi, do you have any additional feedback? Thanks for taking a look again

**Comment by alanocallaghan** [comment]
Are you planning on using the distance matrix stored as an attribute?

Otherwise looks good to me, just update `DESCRIPTION` and describe the change in `inst/NEWS.Rd`

**Comment by FelixErnst** [comment]
Yes. for some types, the distance matrix can be used adjacent to the MDS result, so I didn't want to let it go to waste. However, usually thats not required, so the default is set to `FALSE`

**Comment by FelixErnst** [comment]
sorry about these weird line break changes. I am not sure where they are coming from, since I didn't touch these lines

**Comment by alanocallaghan** [comment]
Seems like a text editor setting (trim trailing whitespace on save in sublime for example). I don't mind it

**Pull request merged**